Thanks, Christophe for the PR.

I would love to see support for HG in `towncrier`

In order to merge this PR, it's critical to have all the tests succesfully executed and all new code to be covered by tests.

We still support Python 3.8 and Python 3.9

We can remove support for 3.8 ... but in a separate PR.

And if supporting Python 3.9 is too much effort, we can also consider removing support for Python 3.9

mypy checkes are part of our standard tests

```
src/towncrier/_hg.py:18: error: Function is missing a return type annotation  [no-untyped-def]
src/towncrier/_hg.py:42: error: Returning Any from function declared to return "bool"  [no-any-return]
src/towncrier/_vcs.py:6: error: Function is missing a return type annotation  [no-untyped-def]
src/towncrier/_vcs.py:26: error: Returning Any from function declared to return "str | None"  [no-any-return]
src/towncrier/_vcs.py:30: error: Returning Any from function declared to return "None"  [no-any-return]
src/towncrier/_vcs.py:34: error: Returning Any from function declared to return "None"  [no-any-return]
src/towncrier/_vcs.py:38: error: Returning Any from function declared to return "list[str]"  [no-any-return]
src/towncrier/_vcs.py:44: error: Returning Any from function declared to return "list[str]"  [no-any-return]
Found 8 errors in 2 files (checked [19](https://github.com/twisted/towncrier/actions/runs/14451901365/job/40526351238?pr=699#step:7:20) source files)
```

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -164,6 +164,7 @@ skip_covered = true
 exclude_lines = [
     "pragma: no cover",
     "if TYPE_CHECKING:",
+    ": \\.\\.\\.$",
 ]
 omit = [
     "src/towncrier/__main__.py",

--- a/src/towncrier/_git.py
+++ b/src/towncrier/_git.py
@@ -5,7 +5,7 @@ from __future__ import annotations
 
 import os
 
-from subprocess import STDOUT, CalledProcessError, call, check_output
+from subprocess import STDOUT, call, check_output
 from typing import Container
 from warnings import warn
 
@@ -29,13 +29,9 @@ def remove_files(fragment_filenames: list[str]) -> None:
         return
 
     # Filter out files that are unknown to git
-    try:
-        git_fragments = check_output(
-            ["git", "ls-files"] + fragment_filenames, encoding="utf-8"
-        ).split("\n")
-    except CalledProcessError:
-        # we may not be in a git repository
-        git_fragments = []
+    git_fragments = check_output(
+        ["git", "ls-files"] + fragment_filenames, encoding="utf-8"
+    ).split("\n")
 
     git_fragments = [os.path.abspath(f) for f in git_fragments if os.path.isfile(f)]
     call(["git", "rm", "--quiet", "--force"] + git_fragments)

--- a/src/towncrier/_hg.py
+++ b/src/towncrier/_hg.py
@@ -5,7 +5,7 @@ from __future__ import annotations
 
 import os
 
-from subprocess import STDOUT, CalledProcessError, call, check_output
+from subprocess import STDOUT, call, check_output
 from typing import Container
 
 
@@ -15,49 +15,22 @@ def get_default_compare_branch(branches: Container[str]) -> str | None:
     return None
 
 
-def _topic_enabled(directory: str) -> bool:
-    for e in (
-        check_output(
-            ["hg", "config"],
-            cwd=directory,
-            encoding="utf-8",
-        )
-        .strip()
-        .splitlines()
-    ):
-        if (
-            e.startswith("extensions.topic")
-            and e.split("=")[0].strip() == "extensions.topic"
-        ):
-            return True
-    return False
-
-
-_has_topics_cache = {}
-
-
-def has_topics(directory: str) -> bool:
-    if directory not in _has_topics_cache:
-        _has_topics_cache[directory] = _topic_enabled(directory)
-
-    return _has_topics_cache[directory] is True
-
-
 def remove_files(fragment_filenames: list[str]) -> None:
     if not fragment_filenames:
         return
 
-    # Filter out files that are unknown to git
-    try:
-        hg_fragments = check_output(
-            ["hg", "files"] + fragment_filenames, encoding="utf-8"
-        ).split("\n")
-    except CalledProcessError:
-        # we may not be in a git repository
-        hg_fragments = []
+    # Filter out files that are unknown to mercurial
+    hg_fragments = (
+        check_output(["hg", "files"] + fragment_filenames, encoding="utf-8")
+        .strip()
+        .split("\n")
+    )
 
     hg_fragments = [os.path.abspath(f) for f in hg_fragments if os.path.isfile(f)]
-    call(["hg", "rm", "--force"] + hg_fragments)
+    fragment_filenames = [
+        os.path.abspath(f) for f in fragment_filenames if os.path.isfile(f)
+    ]
+    call(["hg", "rm", "--force"] + hg_fragments, encoding="utf-8")
     unknown_fragments = set(fragment_filenames) - set(hg_fragments)
     for unknown_fragment in unknown_fragments:
         os.remove(unknown_fragment)
@@ -69,19 +42,10 @@ def stage_newsfile(directory: str, filename: str) -> None:
 
 def get_remote_branches(base_directory: str) -> list[str]:
     branches = check_output(
-        ["hg", "branch"],
+        ["hg", "branches", "--template", "{branch}\n"],
         cwd=base_directory,
         encoding="utf-8",
-        stderr=STDOUT,
     ).splitlines()
-
-    if has_topics(base_directory):
-        branches += check_output(
-            ["hg", "topic", "--template", "{topic}"],
-            cwd=base_directory,
-            encoding="utf-8",
-            stderr=STDOUT,
-        ).splitlines()
 
     return branches
 

--- a/src/towncrier/_hg.py
+++ b/src/towncrier/_hg.py
@@ -15,7 +15,7 @@ def get_default_compare_branch(branches: Container[str]) -> str | None:
     return None
 
 
-def _topic_enabled(directory: str):
+def _topic_enabled(directory: str) -> bool:
     for e in (
         check_output(
             ["hg", "config"],
@@ -39,7 +39,8 @@ _has_topics_cache = {}
 def has_topics(directory: str) -> bool:
     if directory not in _has_topics_cache:
         _has_topics_cache[directory] = _topic_enabled(directory)
-    return _has_topics_cache[directory]
+
+    return _has_topics_cache[directory] is True
 
 
 def remove_files(fragment_filenames: list[str]) -> None:

--- a/src/towncrier/_novcs.py
+++ b/src/towncrier/_novcs.py
@@ -1,3 +1,5 @@
+from __future__ import annotations
+
 import os
 
 from typing import Container

--- a/src/towncrier/_vcs.py
+++ b/src/towncrier/_vcs.py
@@ -1,9 +1,22 @@
+from __future__ import annotations
+
 import os
 
-from typing import Container
+from typing import Container, Protocol
 
 
-def _get_mod(base_directory: str):
+class VCSMod(Protocol):
+    def get_default_compare_branch(self, branches: Container[str]) -> str | None: ...
+    def remove_files(self, fragment_filenames: list[str]) -> None: ...
+    def stage_newsfile(self, directory: str, filename: str) -> None: ...
+    def get_remote_branches(self, base_directory: str) -> list[str]: ...
+
+    def list_changed_files_compared_to_branch(
+        self, base_directory: str, compare_with: str, include_staged: bool
+    ) -> list[str]: ...
+
+
+def _get_mod(base_directory: str) -> VCSMod:
     if os.path.exists(os.path.join(base_directory, ".git")):
         from . import _git
 
@@ -11,11 +24,15 @@ def _get_mod(base_directory: str):
     elif os.path.exists(os.path.join(base_directory, ".hg")):
         from . import _hg
 
-        return _hg
-    else:
+        hg: VCSMod = _hg
+
+        return hg
+    elif base_directory == "/":
         from . import _novcs
 
         return _novcs
+    else:
+        return _get_mod(os.path.dirname(base_directory))
 
 
 def get_default_compare_branch(

--- a/src/towncrier/test/test_check.py
+++ b/src/towncrier/test/test_check.py
@@ -3,7 +3,6 @@
 
 import os
 import os.path
-import warnings
 
 from pathlib import Path
 from subprocess import check_call
@@ -11,7 +10,6 @@ from subprocess import check_call
 from click.testing import CliRunner
 from twisted.trial.unittest import TestCase
 
-from towncrier import check
 from towncrier.build import _main as towncrier_build
 from towncrier.check import _main as towncrier_check
 
@@ -400,24 +398,6 @@ class TestChecker(TestCase):
 
         self.assertEqual(1, result.exit_code)
         self.assertEqual("Could not detect default branch. Aborting.\n", result.output)
-
-    def test_get_default_compare_branch_main(self):
-        """
-        If there's a remote branch origin/main, prefer it over everything else.
-        """
-        branch = check._get_default_compare_branch(["origin/master", "origin/main"])
-
-        self.assertEqual("origin/main", branch)
-
-    def test_get_default_compare_branch_fallback(self):
-        """
-        If there's origin/master and no main, use it and warn about it.
-        """
-        with warnings.catch_warnings(record=True) as w:
-            branch = check._get_default_compare_branch(["origin/master", "origin/foo"])
-
-        self.assertEqual("origin/master", branch)
-        self.assertTrue(w[0].message.args[0].startswith('Using "origin/master'))
 
     @with_isolated_runner
     def test_in_different_dir_with_nondefault_newsfragments_directory(self, runner):

--- a/src/towncrier/test/test_git.py
+++ b/src/towncrier/test/test_git.py
@@ -2,6 +2,8 @@
 # See LICENSE for details.
 
 
+import warnings
+
 from twisted.trial.unittest import TestCase
 
 from towncrier import _git
@@ -13,3 +15,21 @@ class TestGit(TestCase):
         If remove_files gets an empty list, it returns gracefully.
         """
         _git.remove_files([])
+
+    def test_get_default_compare_branch_main(self):
+        """
+        If there's a remote branch origin/main, prefer it over everything else.
+        """
+        branch = _git.get_default_compare_branch(["origin/master", "origin/main"])
+
+        self.assertEqual("origin/main", branch)
+
+    def test_get_default_compare_branch_fallback(self):
+        """
+        If there's origin/master and no main, use it and warn about it.
+        """
+        with warnings.catch_warnings(record=True) as w:
+            branch = _git.get_default_compare_branch(["origin/master", "origin/foo"])
+
+        self.assertEqual("origin/master", branch)
+        self.assertTrue(w[0].message.args[0].startswith('Using "origin/master'))

--- a/src/towncrier/test/test_hg.py
+++ b/src/towncrier/test/test_hg.py
@@ -2,14 +2,14 @@
 # See LICENSE for details.
 
 
-from twisted.trial.unittest import TestCase
-
 from pathlib import Path
 from subprocess import check_call
 
+from click.testing import CliRunner
+from twisted.trial.unittest import TestCase
+
 from towncrier import _hg, _vcs
 
-from click.testing import CliRunner
 from .helpers import setup_simple_project, write
 
 

--- a/src/towncrier/test/test_hg.py
+++ b/src/towncrier/test/test_hg.py
@@ -4,12 +4,76 @@
 
 from twisted.trial.unittest import TestCase
 
-from towncrier import _hg
+from pathlib import Path
+from subprocess import check_call
+
+from towncrier import _hg, _vcs
+
+from click.testing import CliRunner
+from .helpers import setup_simple_project, write
+
+
+def create_project(
+    pyproject_path="pyproject.toml", main_branch="main", extra_config=""
+):
+    setup_simple_project(pyproject_path=pyproject_path, extra_config=extra_config)
+    Path("foo/newsfragments/123.feature").write_text("Adds levitation")
+    initial_commit(branch=main_branch)
+    check_call(["hg", "branch", "otherbranch"])
+
+
+def initial_commit(branch="main"):
+    """
+    Create a mercurial repo, configure it and make an initial commit
+
+    There must be uncommitted changes
+    """
+    check_call(["hg", "init", "."])
+    check_call(["hg", "branch", branch])
+    commit("Initial Commit")
+
+
+def commit(message):
+    """Commit the repo in the current working directory
+
+    There must be uncommitted changes
+    """
+    check_call(["hg", "addremove", "."])
+    check_call(["hg", "commit", "-m", message])
 
 
 class TestHg(TestCase):
+    def test_get_default_compare_branch(self):
+        assert _hg.get_default_compare_branch(["main", "default"]) == "default"
+        assert _hg.get_default_compare_branch(["main", "a_topic"]) is None
+
     def test_empty_remove(self):
         """
         If remove_files gets an empty list, it returns gracefully.
         """
         _hg.remove_files([])
+
+    def test_hg(self):
+        runner = CliRunner()
+        with runner.isolated_filesystem():
+            create_project("pyproject.toml")
+
+            # make sure _vcs._get_mod properly detects a mercurial repo
+            self.assertEqual(_hg, _vcs._get_mod("."))
+
+            write("changes/000.misc.rst", "some change")
+
+            _hg.stage_newsfile(".", "changes/000.misc.rst")
+
+            commit("commit 1")
+
+            branches = sorted(_hg.get_remote_branches("."))
+            self.assertEqual(["main", "otherbranch"], branches)
+
+            self.assertEqual(
+                ["changes/000.misc.rst"],
+                _hg.list_changed_files_compared_to_branch(".", "main", False),
+            )
+
+            write("changes/001.misc.rst", "some change")
+            _hg.remove_files(["changes/000.misc.rst", "changes/001.misc.rst"])

--- a/src/towncrier/test/test_novcs.py
+++ b/src/towncrier/test/test_novcs.py
@@ -1,0 +1,22 @@
+from twisted.trial.unittest import TestCase
+
+from towncrier import _novcs
+
+
+class TestNOVCS(TestCase):
+    def test_get_default_compare_branch(self):
+        self.assertEqual(None, _novcs.get_default_compare_branch([]))
+
+    def test_empty_remove(self):
+        """
+        If remove_files gets an empty list, it returns gracefully.
+        """
+        _novcs.remove_files([])
+
+    def test_get_remote_branches(self):
+        self.assertEqual([], _novcs.get_remote_branches("."))
+
+    def test_list_changed_files_compared_to_branch(self):
+        self.assertEqual(
+            [], _novcs.list_changed_files_compared_to_branch(".", "main", False)
+        )

